packages/browser-wallet/src/background/recovery.ts

```diff
@@ -0,0 +1,102 @@
+import {
+    createIdentityRecoveryRequest,
+    CryptographicParameters,
+    IdentityObjectV1,
+    IdentityRecoveryRequestInput,
+    Network as NetworkString,
+    Versioned,
+} from '@concordium/web-sdk';
+import { ExtensionMessageHandler } from '@concordium/browser-wallet-message-hub';
+import { IdentityProviderIdentityStatus, IdentityTokenContainer, sleep } from 'wallet-common-helpers';
+import { Identity, CreationStatus, IdentityProvider } from '@shared/storage/types';
+import { BackgroundResponseStatus } from '@shared/utils/types';
+import { addIdentity } from './update';
+
+// TODO: increase max empty
+// How many empty identityIndices are allowed before stopping
+const maxEmpty = 3;
+// Milliseconds to wait, if retrievelUrl is still pending
+const sleepInterval = 5000;
+
+export type Payload = {
+    providers: IdentityProvider[];
+    globalContext: CryptographicParameters;
+    seedAsHex: string;
+    net: NetworkString;
+};
+
+/**
+ * Polls the provided location until a valid identity object is available, or that an error is returned.
+ * If an error occurs or is returned from the url, this return undefined
+ */
+async function getIdentityObject(url: string): Promise<Versioned<IdentityObjectV1> | undefined> {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+        const response = (await (await fetch(url)).json()) as IdentityTokenContainer;
+        if (response.status === IdentityProviderIdentityStatus.Error) {
+            return undefined;
+        }
+        if (response.status === IdentityProviderIdentityStatus.Done) {
+            return response.token.identityObject;
+        }
+        await sleep(sleepInterval);
+    }
+}
+
+function getRecoverUrl(inputs: Omit<IdentityRecoveryRequestInput, 'timestamp' | 'ipInfo'>, provider: IdentityProvider) {
+    const timestamp = Math.floor(Date.now() / 1000);
+    const idRecoveryRequest = createIdentityRecoveryRequest({ ...inputs, timestamp, ipInfo: provider.ipInfo });
+
+    const searchParams = new URLSearchParams({
+        state: JSON.stringify({ idRecoveryRequest }),
+    });
+    return `${provider.metadata.recoveryStart}?${searchParams.toString()}`;
+}
+
+async function performRecovery({ providers, ...recoveryInputs }: Payload) {
+    // const identities = await storedIdentities.get();
+    // let nextId = identities ? identities.length + 1 : 0;
+    let nextId = 0;
+    const identitiesToAdd: Identity[] = [];
+
+    for (const provider of providers) {
+        // TODO: remove this when providers have been fixed
```

Comment:
```suggestion
        // TODO: Is required because some identity providers do not have a recoveryStart value. This is an error and should be fixed in the wallet proxy. At that point this can be safely removed.
```